Creating ADR for versioning and branching strategy (#4998)
dbt-labs/dbt-core: docs/arch/adr-003-versioning-branching-strategy.md

We need a branching strategy to support the goals of the versioning strategy. We do not want to have to stop development on `main` as we await a release, nor do we want every change in `main` published in every release. Using branches, we can decide which changes go into which release.

## Common Terminology
* Release version: refers to the numeric value of the release (eg. `1.0.1`, `0.21.0`) There are major, minor, and patch release versions that map to standard [semantic versioning practices](https://semver.org/#semantic-versioning-200).
* Release phase: refers to the stability of the release (eg. beta, rc, final). 

## Requirements
The following are requirements that must be met for releasing dbt:

1. We must have major, minor, and patch releases.
1. We must have release phases that have the goal of garnering early feedback from the Community.
1. We must have release phases that aim to identify bugs and test for stability.
1. We must have release phases that are stable and of production quality.
1. We must have the ability to isolate changes from different versions and phases to release.

## Branching Strategy Options

> The 2 branching strategies we are considering using going forward are:

Reviewer: I think this should be more like "alternatives we considered" since we're continuing with our current strategy after talking about it.

Author: What about "Branching Strategies Considered"? In ADR's I've seen, it's more about listing out the options in the first half of the doc and then the decisions made and why in the second half. I like having the 2 strategies juxtaposed here to see the pros and cons. I do call out the current strategy so we remember going forward that we kept with the existing position but this time we knows the reason why it was chosen

* Confusing for backporting. We will need to keep creating new labels and making sure we remove the old labels because we don't want to accidently backport to the wrong branch.

### Minor release branches (current strategy)
#### Pros:
* Adapter CI testing will continue to work with this strategy and makes it less complicated overall. Adapters and `dbt-core` need to be on the same minor version but can be on different patch versions. The current CI workflows verify that the changes slated for a new `1.1` patch are compatible across all repos by checking out the `1.1.latest` branch of each repo. The 1:1 mapping of adapter branches to `dbt-core` branches makes this easy.
* Easy backporting. There is no need to figure out the exact branch that you need to backport to and accidently backporting to an older, stale branch.
#### Cons:
* Branches contain more than 1 release which is confusing. Without a 1:1 mapping between branch and release, you must rely on tags instead to know where in the branch the release was cut.

> * Releasing from the same branch multiple times can lead to confusion as to which version we are releasing and actually stops us from releasing the wrong version.

Reviewer: > and actually stops us from releasing the wrong version.

I think I know what you're getting at, but the wording is confusing to me here. Maybe it should be a separate sentence.

Author: I see what you mean about the wording. I took another go at clarifying that

Reviewer: the specifics make it much clearer. thanks!

- Branches: the branches where these release phases are present (e.g. betas should only exist on the `main` branch)
- Applicable Release Versions: the release versions where the release phase is applicable (i.e. we will only have alphas for major and minor versions, we will not have alphas for patch versions)
- Expectations: the stability of the code changes in the release
</details>

| phase | Released? | Branches | Applicable Release Versions | Expectations |
| ---- | --------- | -------- | ---------------- | ------------ |
| Alpha | No | `main` | Major / Minor | Experimental |
| Beta | Yes | `main` | Major / Minor | Experimental |
| RC | Yes | Release branch | Major / Minor / Patch | Pre-production |
| Final | Yes | Release branch | Major / Minor / Patch | Production |

### Branching Strategy
Based off of the release version and expectations, the version phase's branching strategy can be determined. 
 * Alpha and Beta versions are experimental, and therefore exist only on the `main` branch, where all our changes reside
 * Alpha versions are never released. They signify that development is underway, though the changes they contain have not been released in any form

>  * An RC version denotes a more stable state. A release branch will be created for RC releases to limit the changes a release contains

Reviewer: we should probably use the term "release candidate" here even though it's implied.

\* Disclaimer: the how and when to apply changes to a release branch will be covered in another upcoming ADR.

1. If a release branch is not yet cut: This is the scenario where we start to develop a feature for a future version before we have cut a release branch for the upcoming version (eg. we want to work on a feature for `1.2.0` but we haven't cut the release branch for `1.1.0` yet). 
    1. Is this dead code? If so, then commiting the changes to `main` is acceptable. There must be 0 risk of users hitting this so if unsure, please uses the alternative option.
    1. Use a feature branch until the release branch is cut. Then when it makes sense, merge the feature branch into `main` and continue development.

   ** Sometimes feature flags are used by teams to also achieve this goal. In our case, feature flags aren't controlled by the team and instead dependent of a user to toggle them on/off. Core uses feature flags instead for offering experimental functionality for users to try and test out. This is different from hiding features from going into certain releases.

## Status

> In Progress

Author: I always save this for the last commit :-)

# Release Versioning / Branching Strategy

## Context

> With `dbt` ever evolving and progressing, we continue to ship and release versions on a regular basis. As we release `dbt`, versioning and branching become key to identifying what changes for users and conveying the expectations of those changes. We currently push all our changes to `main` which is the latest version of our code. We don't always want to release the latest version of our code though. We want to be able to isolate changes to go into particular releases based on urgency (e.g. bug fixes) and stability (e.g. new features). 

Reviewer: > As we release `dbt`, versioning and branching become key to identifying what changes for users and conveying the expectations of those changes.

This sentence uses "changes" as both a noun and a verb in the same sentence and confused me. Maybe use "is different" for the verb?

We already have an established versioning strategy when it comes to major/minor/patch releases. The following is our approach as to what to expect from each release version:

> * Major: This signifies that breaking changes to users that they must react to are included in this release.

Reviewer 2: ```suggestion
* Major: This signifies that breaking changes which require user action are included in this release.
```

> * Minor version: This signifies that new features are included in this release.

Reviewer: are we making the guarantee that there will be no breaking changes here either?

Author: In minor releases? Correct, no breaking changes. CC @jtcohen6 to double confirm 😄 b/c I know we did put a breaking changelog in the other day...

Reviewer 3: No breaking changes for project code / end users of dbt!!!

We reserve the right to make breaking changes ONLY for:
- the adapter interface (Python methods), relevant to maintainers of adapter plugins
- metadata interfaces — new versions of JSON artifact versions, new version of structured logging

We should still aim to avoid those changes whenever possible. When they do happen, we should communicate them proactively and clearly, and call out _exactly_ who they're for. Probably adapter plugin maintainers. E.g.
https://github.com/dbt-labs/dbt-core/blob/9c233f27ab280f36cc5b3007d0b85a0f9d3ad13a/.changes/1.1.0/Breaking%20Changes-20220316-031311.yaml#L3
https://github.com/dbt-labs/dbt-core/blob/9c233f27ab280f36cc5b3007d0b85a0f9d3ad13a/.changes/1.1.0/Breaking%20Changes-20220412-152450.yaml#L2

* Patch version: This signifies that bug and security fixes are included in this release.

> We need a versioning strategy that allows us to to convey to users the confidence and stability of code they are installing. We want a versioning strategy that allows users to try out new changes and experiment with them to get early feedback. We want to make sure users can also get a preview of new features we are working on so that they can incorporate those changes as soon as possible to their own work. 

Reviewer 2: ```suggestion
We need a versioning strategy that allows us to to convey to users the confidence and stability of code they are installing. We want a versioning strategy that allows users to try out new changes, and experiment with them to get early feedback. We want to make sure users can also get a preview of new features we are working on, so they can incorporate these changes into their work as soon as possible. 
```

> We need a branching strategy to support the goals of the versioning strategy. We do not want to have to stop development to `main` as we await a release. We also don't want to have everything in `main` to go into every release. Because of this we can leverage branches to ensure what changes are going into what release and be deliberate about those changes. 

Reviewer 2: ```suggestion
We need a branching strategy to support the goals of the versioning strategy. We do not want to have to stop development on `main` as we await a release, nor do we want every change in `main` published in every release. Using branches, we can ensure which changes go into which release.
```

## Requirements

> The following are requirements that must be met for releasing dbt:

Reviewer 4: I think we need to make further distinctions here for the target of the release (pypi, docker, github, etc) because not all of them fit into this list (docker images specifically)

Reviewer 4: Edit:  Is target what you meant by `type`?

Author: @iknox-fa I'm changing type to phase. I used the wrong term.

Also can you elaborate on what you would like to see in requirements for the release targets? Do you mean that Docker always releases from the `main` branch? I was thinking of creating a new doc specifically for the release targets but do you think that will impact the statements made in the requirements here?

Author: I added another requirement bullet point to define this. @jtcohen6 would you agree with my last point of what "released" means to us?

1. We must have major, minor, and patch releases.
1. We must have release types that have the goal of garnering early feedback from the Community.

> 1. We must have release types that aim to identify bugs and test for stability.

Reviewer: What is a release type

Author: I was struggling for a good word here. I can give examples (beta, rc, etc.). Any thought on a better term? I didn't want to use version since we are using version in the context of major, minor, and patch.

Reviewer: release "stage" maybe?

Author: Ok updating to that. Thanks for the idea!

Reviewer 5: [PEP 440](https://peps.python.org/pep-0440/#pre-releases) uses the term "phase" to describe options like beta, rc, etc within the pre-release "segment".

Reviewer: I also like "phase."

> 1. We must have release types that are stable and production quality.

Reviewer 2: ```suggestion
1. We must have release types that are stable and of production quality.
```

1. We must have the ability to isolate changes from different versions based on the type of release.

## Decisions

### Version Types
<details>
<summary> Header Explanations: </summary>

> - Type: type of version that is occurring (i.e. a beta vs final)

Reviewer 2: ```suggestion
- Type: the type of the version (e.g. beta, final)
```

- Released?: does this type of version need to be released?

> - Branches: the branches where these version types are present (i.e. we will only find beta version numbers on the `main` branch)

Reviewer 2: ```suggestion
- Branches: the branches where these version types are present (e.g. beta version numbers should only exist on the `main` branch)
```

- Expectations: the stability of the code changes in the release
</details>

| Type | Released? | Branches | Release Versions | Expectations |
| ---- | --------- | -------- | ---------------- | ------------ |
| Alpha | No | `main` | Major / Minor | Experimental |
| Beta | Yes | `main` | Major / Minor | Experimental |
| RC | Yes | Release branch | Major / Minor / Patch | Pre-production |
| Final | Yes | Release branch | Major / Minor / Patch | Production |

### Branching Strategy
Based off of the release version and expectations, the version type's branching strategy can be determined. 
 * Alpha and Beta versions are experimental, and therefore exist only on the `main` branch, where all our changes reside
 * Alpha versions are never released. They signify that development is underway, though the changes they contain have not been released in any form
 * An RC version denotes a more stable state. A release branch will be created for RC releases to limit the changes a release contains
 * Final versions are stable and tested. Only verified changes will go into these releases, which are inherited from the corresponding RC version

>  * A release branch will exist for each unique major or minor version, to be named accordingly: `<major>.<minor>.latest` (e.g. `1.0.latest`)

Reviewer 4: I think we should consider creating a release branch for every release down to the patch level.  There have been at least few cases where that would have made life much easier.

Reviewer: I have also liked working on projects that have a 1:1 relationship between releases and branches.

Reviewer 6: @iknox-fa can you explain what we gain by doing it this way?  It seems to me it adds complexity of what branch to merge to.  It also adds complexity to the backport action and our changelog generation.

Reviewer: @emmyoop the way I've done it, you cut a branch for the release then never touch it again. No more figuring out what commit a release was released from and rolling back your local git to investigate bugs.

Author: @iknox-fa or @nathaniel-may have you then just created patch branches off of the initial minor version branch? So for example: 
For the beta, I create the `1.1.0` branch off of `main`
Then for `1.1.1` patch, I create the `1.1.1` branch off of the `1.1.0` branch?
What about for RCs? Do they get their own branch as well?

Reviewer: @leahwicz exactly like that. I'm used to all releases having a branch including RCs.

Author: @iknox-fa @nathaniel-may we tag the commit that is released. Are tags not a good way to map what we released to what version? Relying solely on a branch for having only the commits we released actually scare me since in theory you can add more commits to them after a release (not that it would be intentional) but a tag is a definitive end commit.

Reviewer: tags do work for this mapping (I wrote more about it in our slack thread). But the same worry exists with our current strategy because we could edit the history of the branch. It might be a little harder to do by accident, but we already live with that possibility. In both situations we would have branch protections to limit the surface area of that risk.

>  ![Branching Strategy](images/ReleasingBranchStrategy.png)

Reviewer: to this point, maybe this doc can explicitly describe what the plan is for a commit we do _not_ want to go out in the next release is?

>  ![Branching Strategy](images/ReleasingBranchStrategy.png)

Author: Historically, everything in `main` has gone into the release branch. We have never started (not saying this can't happen in the future) working on something that was slated for the next release before the release branch was cut for the upcoming release. In most cases, I would think this would most likely happen for a large feature and in this case we should leverage feature branches which I can add to the doc to make it clear on how we should handle something like this if it would arise

>  ![Branching Strategy](images/ReleasingBranchStrategy.png)

Reviewer: yeah that works. just something we can have to start with and iterate on.

>  ![Branching Strategy](images/ReleasingBranchStrategy.png)

Author: I added a section about this so let me know if that helps clarify things

>  ![Branching Strategy](images/ReleasingBranchStrategy.png)

Reviewer: it does! thanks.

>  * Alpha and Beta are experimental and therefore exist on `main` where all our changes reside. 
>  * Alpha is never released but instead signifies that development is happening but the changes during that period have never been released in any form.
>  * RC is moving to a stable state. Therefore, a release branch will be created for RC releases to start limiting the changes happening for that release.
>  * Final is a stable, tested version. Only verified changes will go into this release which is inherited from the RC so a release branch is also required.
>  * A release branch will exist for each unique major, minor version release and be named accordingly: `<major>.<minor>.latest` (ex `1.0.latest`)

Reviewer 2: ```suggestion
 * Alpha and Beta versions are experimental, and therefore exist only on the `main` branch, where all our changes reside
 * Alpha versions are never released. They signify that development is underway, though they changes they contain have not been released in any form
 * An RC version denotes a more stable state. A release branch will be created for RC releases to limit the changes contained in a release
 * Final versions are stable and tested. Only verified changes will go into these releases which are inherited from the corresponding RC version
 * A release branch will exist for each unique major or minor version, to be named accordingly: `<major>.<minor>.latest` (e.g. `1.0.latest`)
```

## Status
In Progress

## Consequences

> This doesn't drastically change our current versioning and branching strategy but instead documents what we have been doing.
> 
> The only major change is we are now incorporating an Alpha version which we previously didn't use at all. The consequences of moving to use an Alpha are to make developing across multiple repos easier. We want `dbt-core` and adapters depend on minor versions. If we don't have a way to keep the `main` branches in sync with one another, then integration tests will start failing. It is also confusing when the `main` branch shows it's an RC or Final version when we do not release the version from there. This gives more clarity as to what versions live where.

Reviewer 2: ```suggestion
This doesn't drastically change our current versioning or branching strategies, but documents our existing flows.

The only real change here is the addition of an Alpha release, the intent of which is to make developing across multiple repos easier. `dbt-core` and adapters should depend on minor versions. If we don't have a way to keep the `main` branches in sync with one another, integration tests will start to fail. It is also confusing when the `main` branch is marked as an RC or Final version when we never release from `main`. This change provides visibility around which versions live where.
```

## Outside Scope
The following are topics that are outside the scope of this document and will be addressed in their own ADR:

> * Hotfixes and how branching works for them
> * How and why we bump release versions right before a release.

Reviewer 2: ```suggestion
* Hotfixes and their branching strategies
* How and why we bump release versions right before a release.
```

 ![Branching Strategy](images/ReleasingBranchStrategy.png)

### Changes for Specific Release Versions
There will be times that we will want to target changes to go into specific release versions. Below outlines 2 different scenarios in which this can occur and how we will approach them.

1. If a release branch already exists: This is the most common scenario where we want to add a change to an upcoming release and the release branch has already been cut. In this instance, developers will commit their changes to `main` first. Those changes (or some form of them) will then be applied to the release branch in question so that they can be included in the specific release version. 
\* Disclaimer: the how and when to apply changes to a release branch will be covered in another upcoming ADR.

> 1. If a release branch is not yet cut: This is the scenario where we start to develop a feature for a future version before we have cut a release branch for the upcoming version (eg. we want to work on a feature for `1.2.0` but we haven't cut the release branch for `1.1.0` yet). 

Reviewer: all the number bullets are "1"

Author: In markdown, you just use `1.` everywhere and when markdown renders, it uses the correct numbers. If you look at this in the rich diff view for the PR, it shows correctly for me. Let me know if that isn't the case for you
